msal/src/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs

```diff
@@ -195,9 +195,10 @@ private void SaveTokenResponseToCache()
 
             ClientInfo fromServer = null;
 
-            if (!AuthenticationRequestParameters.IsClientCredentialRequest)
+            if (!AuthenticationRequestParameters.IsClientCredentialRequest && AuthenticationRequestParameters.Authority.AuthorityType != Core.Instance.AuthorityType.Adfs)
             {
                 //client_info is not returned from client credential flows because there is no user present.
+                //Adfs does not return client_info
```

Comment:
Caching worked in my tested scenarios. I populate the HomeAccountID with the subjec tin the ID token and leave the tenantId null

Reply:
The problem is that an ADALv3 user will also have an empty client_info ... This will be a pretty complex test case...

---
In reply to: [223772308](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1293#discussion_r223772308) [](ancestors = 223772308)